plugins/auth-backend/src/providers/gitlab/provider.ts

```diff
@@ -47,29 +57,63 @@ type PrivateInfo = {
 
 export type GitlabAuthProviderOptions = OAuthProviderOptions & {
   baseUrl: string;
+  signInResolver?: SignInResolver<OAuthResult>;
+  authHandler: AuthHandler<OAuthResult>;
+  tokenIssuer: TokenIssuer;
+  catalogIdentityClient: CatalogIdentityClient;
+  logger: Logger;
 };
 
-function transformProfile(fullProfile: FullProfile) {
-  const profile = makeProfileInfo({
-    ...fullProfile,
-    photos: [
-      ...(fullProfile.photos ?? []),
-      ...(fullProfile.avatarUrl ? [{ value: fullProfile.avatarUrl }] : []),
-    ],
+export const extractGitLabUserId = (profile: ProfileInfo): string => {
+  return profile.email?.split('@')[0] as string;
+};
+
+export const gitlabDefaultSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { result } = info;
+
+  const id = result.fullProfile.username || result.fullProfile.id;
+
+  const token = await ctx.tokenIssuer.issueToken({
+    claims: { sub: id, ent: [`user:default/${id}`] },
   });
 
-  let id = fullProfile.id;
-  if (profile.email) {
-    id = profile.email.split('@')[0];
-  }
+  return { id, token };
+};
 
-  return { id, profile };
-}
+export const gitlabDefaultAuthHandler: AuthHandler<OAuthResult> = async ({
+  fullProfile,
+  params,
+}) => {
+  fullProfile.photos = [
+    ...(fullProfile.photos ?? []),
+    ...((fullProfile as FullProfile).avatarUrl
+      ? [{ value: (fullProfile as FullProfile).avatarUrl as string }]
+      : []),
+  ];
```

Comment:
Tbh I think the cleanest thing here would be to move this logic into the common `makeProfileInfo` helper. That is if there is an `avatarUrl` and no photos, use it as the profile picture.